### PR TITLE
Add unit test base files. Add namespace naming check.

### PR DESCRIPTION
**Pull request by ZdravkoD**
Made small changes on the **add_new_check.py**.
Added new check for namespace naming convention.
Added new python script **check_clang_tidy.py**, used for running tests over a test file.
Added **docs** folder.

### CODE REVIEW

File: run_tests.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+alienv enter o2codechecker/latest GCC-Toolchain/latest; 
+O2codecheck ex.cpp -fix --checks=-*,alice* --
```
Comment (sawenzel) [inline run_tests.sh]: In which directory is this file? I would not put it in the main one.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -2,5 +2,4 @@
 add_executable(testregex testregex.cpp)
 add_test(testregexcorrect testregex)
 
-
 # correctness testing of checks
```
Comment (sawenzel) [inline test/CMakeLists.txt]: We should add the unit test here in the following manner:
`add_test(aliceo2namespace COMMAND ${CMAKE_BINARY_DIR}/tool/O2codecheck --checks=-*,alice* ${CMAKE_SOURCE_DIR}/test/aliceO2-namespace-naming.cpp --)` (or similar ... with your script that finally gives thumbs up or thumbs down for this test).

This will make sure the tests are run when we say `make test`.

File: test/aliceO2-namespace-naming.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
+// RUN: %check_clang_tidy %s aliceO2-namespace-naming %t
+
```
Comment (sawenzel) [inline test/aliceO2-namespace-naming.cpp]: can you add here:
a) a double nested namespace example
b) a `using namespace Foo` example
c) a simple typedef aka `using MyType = Foo::Bar::ClassA`

File: plugin/PluginTidyModule.cpp [hunk 4]
```diff
@@ -28,12 +28,12 @@ class PluginModule : public ClangTidyModule {
 
 } // namespace plugin
 
-// Register the AliceO2TidyModule using this statically initialized variable.
+// Register the PluginO2TidyModule using this statically initialized variable.
 static ClangTidyModuleRegistry::Add<plugin::PluginModule>
-X("aliceO2-module", "Adds AliceO2 specific checks");
+X("pluginO2-module", "Adds PluginO2 specific checks");
```
Comment (sawenzel) [inline plugin/PluginTidyModule.cpp]: changes like here are not related to this PR. It is preferred to have them in a separate commit/PR to better understand development history.

File: test/aliceO2-namespace-naming.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: namespace 'FirstNamespace' does not follow the underscore convention [aliceO2-namespace-naming]
+// CHECK-FIXES: {{^}}void first_namespace::fooFromNamespace(); // just forward declaration{{$}}
+
+void FirstNamespace::fooFromNamespace() // definition
+// CHECK-MESSAGES: :[[@LINE-1]]:6: warning: namespace 'FirstNamespace' does not follow the underscore convention [aliceO2-namespace-naming]
+// CHECK-FIXES: {{^}}void first_namespace::fooFromNamespace() // definition{{$}}
+{
+}
+
+void foo() // has no parent
```
Comment (sawenzel) [inline test/aliceO2-namespace-naming.cpp]: this is not used?

File: check_clang_tidy.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+  print "@@    @@@     @@   @@    @"
+  print "@@  @@   @@   @@  @@    @@@"
+  print "@@  @@   @@   @@@@      @@@"
+  print "@@  @@   @@   @@@@       @"
+  print "@@  @@   @@   @@  @@     "
+  print "@@    @@@     @@   @@    @"
+  print "@@  "
+  print "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@"
+  print "@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@"
+  print ""
```
Comment (sawenzel) [inline check_clang_tidy.py]: consider (also) returning an error code for the process which is the usual way to transmit pass/fail to other processes (ctest)

File: run_tests.sh (hunk 1 above)
Comment (ZdravkoD) [inline run_tests.sh]: My bad. This file was not meant to be uploaded, because it is not done. Its intent was to run all tests and it should be in the **o2codechecker\test\** directory.

I will remove it.

File: test/aliceO2-namespace-naming.cpp [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+// CHECK-MESSAGES: :[[@LINE-1]]:11: warning: namespace 'FirstNamespace' does not follow the underscore convention [aliceO2-namespace-naming]
+// CHECK-FIXES: {{^}}namespace first_namespace{{$}}
+{
+  int a;
+  void fooFromNamespace();
+  
+  void bar()
+  {
+  }
+
```
Comment (sawenzel) [inline test/aliceO2-namespace-naming.cpp]: How do we check correct examples?

File: test/CMakeLists.txt (hunk 2 above)
Comment (ZdravkoD) [inline test/CMakeLists.txt]: Should **check_clang_tidy.py** be used? If so, I should add an error code return statement to the script as you mentioned in your comment to the script. And where is it specified which script will be used for the test?

File: check_clang_tidy.py (hunk 6 above)
Comment (ZdravkoD) [inline check_clang_tidy.py]: By default Python scripts return "0" error code if they don't raise exception.
This script prints "OK!" when there are no exceptions. Thus, the script returns error code = 0.
This script doesn't print "OK!" when there are exceptions. Thus, the script returns error code != 0.

What we can do is to add different error codes at the places where the script fails, so that we could more easily find out where is the exception. Is that what you meant?

File: test/CMakeLists.txt (hunk 2 above)
Comment (sawenzel) [inline test/CMakeLists.txt]: Yes. Just say
`add_test(aliceo2namespace COMMAND check_clang_tidy.py options ${CMAKE_SOURCE_DIR}/test/aliceO2-namespace-naming.cpp)` or the like.

File: test/aliceO2-namespace-naming.cpp (hunk 7 above)
Comment (ZdravkoD) [inline test/aliceO2-namespace-naming.cpp]: Correct examples are checked when you don't write any annotation "CHECK-MESSAGE" or "CHECK-FIXES". Meaning you don't expect to see any changes in that line of code. If a change occurs(meaning the check saw a problem in that line and throws a warning or reworks the line of code), then the test fails by saying that unexpected change is seen.

File: check_clang_tidy.py (hunk 6 above)
Comment (sawenzel) [inline check_clang_tidy.py]: If a non-zero error code is already returned in case of problems, it is probably enough. 
But in my opinion we don't need the "OK" inside the script. It will be uncessary printout at some stage. You can achieve the same effect yourself by checking the return code from the script and eventually make a printout on the command line (using bash, etc.).

File: test/aliceO2-namespace-naming.cpp (hunk 7 above)
Comment (sawenzel) [inline test/aliceO2-namespace-naming.cpp]: ok good!

File: check_clang_tidy.py (hunk 6 above)
Comment (ZdravkoD) [inline check_clang_tidy.py]: Okay, you are right. I will remove the printing from the script.

File: test/CMakeLists.txt (hunk 2 above)
Comment (ZdravkoD) [inline test/CMakeLists.txt]: In the command that you give, is the filename `${CMAKE_SOURCE_DIR}/test/aliceO2-namespace-naming.cpp` supposed to be an argument to the script? If so, then I should add the following line
`add_test(aliceo2namespace COMMAND check_clang_tidy.py ${CMAKE_SOURCE_DIR}/test/aliceO2-namespace-naming.cpp 'aliceO2-namespace-naming')`
Since the script accepts two arguments: 1st the file name, 2nd the name of the check. Right?

Comment (sawenzel) [inline test/CMakeLists.txt]: Sounds right. Later you can also define a cmake macro `add_tidy_test` that saves some typing and just takes the 2 argument: filename + test.

Comment (ZdravkoD) [inline test/CMakeLists.txt]: Actually we can modify the `add_new_check.py` script to add this line automaticaly when the check is created. So we would not have to worry about it.
